File: public/java/061-jump-game.java
/*
 * LeetCode Problem #55: Jump Game
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/jump-game/
 */
import java.util.*;

class Solution {

    // APPROACH 1: Greedy (Forward) | O(n) time | O(1) space
    // EXPLAIN: Track the farthest index reachable; if current index ever exceeds it, return false.
    // WHEN: Use as the optimal solution — single pass, no extra space needed.

    public boolean canJump(int[] nums) {
        int maxReach = 0;
        for (int i = 0; i < nums.length; i++) {
            if (i > maxReach) return false;
            maxReach = Math.max(maxReach, i + nums[i]);
        }
        return true;
    }

    // APPROACH 2: Greedy (Backward) | O(n) time | O(1) space
    // EXPLAIN: Start from the end; shift the goal leftward whenever a position can reach it.
    // WHEN: Elegant alternative — scan right-to-left without any auxiliary structure.

    public boolean canJump_backward(int[] nums) {
        int lastPos = nums.length - 1;
        for (int i = nums.length - 2; i >= 0; i--) {
            if (i + nums[i] >= lastPos) lastPos = i;
        }
        return lastPos == 0;
    }

    // APPROACH 3: DP | O(n²) time | O(n) space
    // EXPLAIN: dp[i] = true if position i is reachable; for each reachable position extend forward.
    // WHEN: Use when you need to know which positions are reachable (not just the final answer).

    public boolean canJump_dp(int[] nums) {
        int n = nums.length;
        boolean[] dp = new boolean[n];
        dp[0] = true;
        for (int i = 0; i < n; i++) {
            if (!dp[i]) continue;
            for (int j = 1; j <= nums[i] && i + j < n; j++) {
                dp[i + j] = true;
            }
        }
        return dp[n - 1];
    }

    // APPROACH 4: BFS | O(n²) time | O(n) space
    // EXPLAIN: Each position is a node; edges connect to all reachable positions; BFS checks last.
    // WHEN: Use for a graph-traversal perspective that generalises to weighted jump games.

    public boolean canJump_bfs(int[] nums) {
        int n = nums.length;
        if (n == 1) return true;
        boolean[] visited = new boolean[n];
        int maxReach = 0;
        for (int i = 0; i <= maxReach && i < n; i++) {
            maxReach = Math.max(maxReach, i + nums[i]);
            if (maxReach >= n - 1) return true;
        }
        return false;
    }

    // APPROACH 5: Recursive Memoization | O(n²) time | O(n) space
    // EXPLAIN: For each position try all jumps recursively; cache results to avoid re-computation.
    // WHEN: Use as a top-down DP demonstration — not recommended over greedy but instructive.

    public boolean canJump_memo(int[] nums) {
        int[] memo = new int[nums.length]; // 0=unknown, 1=true, -1=false
        return helper(nums, 0, memo);
    }

    private boolean helper(int[] nums, int pos, int[] memo) {
        if (pos >= nums.length - 1) return true;
        if (memo[pos] != 0) return memo[pos] == 1;
        for (int jump = 1; jump <= nums[pos]; jump++) {
            if (helper(nums, pos + jump, memo)) {
                memo[pos] = 1;
                return true;
            }
        }
        memo[pos] = -1;
        return false;
    }
}

// Made with Bob

File: public/python/061-jump-game.py
# LeetCode Problem #55: Jump Game
# Difficulty: Medium
# Link: https://leetcode.com/problems/jump-game/

from typing import List
from collections import deque

# ─────────────────────────────────────────────
# APPROACH 1: Greedy (Forward) | O(n) time | O(1) space
# EXPLAIN: Track the furthest index reachable; if current index exceeds it, return False.
# WHEN: Optimal — single pass, no extra memory. The go-to approach.

def canJump_greedy(nums: List[int]) -> bool:
    max_reach = 0
    for i, jump in enumerate(nums):
        if i > max_reach:
            return False
        max_reach = max(max_reach, i + jump)
    return True


# APPROACH 2: Greedy (Backward) | O(n) time | O(1) space
# EXPLAIN: Start from the end; track the leftmost index that can reach it; if 0, return True.
# WHEN: Elegant alternative — scan right-to-left, no extra data structure needed.

def canJump_greedy_backward(nums: List[int]) -> bool:
    last_pos = len(nums) - 1
    for i in range(len(nums) - 2, -1, -1):
        if i + nums[i] >= last_pos:
            last_pos = i
    return last_pos == 0


# APPROACH 3: DP (Bottom-Up) | O(n²) time | O(n) space
# EXPLAIN: dp[i] = True if index i is reachable; for each reachable i extend forward.
# WHEN: When you want to know which positions are reachable, not just the final answer.

def canJump_dp(nums: List[int]) -> bool:
    n = len(nums)
    dp = [False] * n
    dp[n - 1] = True
    for i in range(n - 2, -1, -1):
        furthest = min(i + nums[i], n - 1)
        for j in range(i + 1, furthest + 1):
            if dp[j]:
                dp[i] = True
                break
    return dp[0]


# APPROACH 4: BFS | O(n²) time | O(n) space
# EXPLAIN: Each index is a node; edges lead to all reachable indices; BFS checks last index.
# WHEN: Useful for visualising the problem as a graph shortest-path variant.

def canJump_bfs(nums: List[int]) -> bool:
    n = len(nums)
    if n == 1:
        return True
    visited = {0}
    queue = deque([0])
    while queue:
        i = queue.popleft()
        for step in range(1, nums[i] + 1):
            nxt = i + step
            if nxt >= n - 1:
                return True
            if nxt not in visited:
                visited.add(nxt)
                queue.append(nxt)
    return False


# APPROACH 5: Recursive Memoization | O(n²) time | O(n) space
# EXPLAIN: For each position recursively try all jumps; cache results to avoid recomputation.
# WHEN: Top-down DP style — shows the problem structure; not optimal but educational.

from functools import lru_cache

def canJump_memo(nums: List[int]) -> bool:
    n = len(nums)

    @lru_cache(maxsize=None)
    def dp(pos: int) -> bool:
        if pos >= n - 1:
            return True
        for jump in range(1, nums[pos] + 1):
            if dp(pos + jump):
                return True
        return False

    return dp(0)


# ── quick tests ──────────────────────────────────────────────────────────────
if __name__ == '__main__':
    cases = [
        ([2, 3, 1, 1, 4], True),
        ([3, 2, 1, 0, 4], False),
        ([0], True),
        ([2, 0, 0], True),
        ([1, 0, 1, 0], False),
    ]
    for fn in (canJump_greedy, canJump_greedy_backward, canJump_dp,
               canJump_bfs, canJump_memo):
        for nums, expected in cases:
            result = fn(nums)
            assert result == expected, (
                f'{fn.__name__}({nums}) = {result}, expected {expected}'
            )
    print('All tests passed.')

# Made with Bob
